File: booking_service/src/ui/kafka/accommodation_event_handler.go
package consumer

import (
	"booking_service/core/domain/constant"
	basedto "booking_service/core/domain/dto/kafka/base"
	"booking_service/core/domain/dto/request"
	"booking_service/core/service"
	"context"
	"encoding/json"
	"github.com/golibs-starter/golib/log"
	"time"
)

type AccommodationEventHandler struct {
	accService service.IAccommodationService
}

func (a *AccommodationEventHandler) HandleMessage(topic string, key, value []byte) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	var domainEventDto basedto.DomainEventDto
	if err := json.Unmarshal(value, &domainEventDto); err != nil {
		log.Error(ctx, "error unmarshal domain event dto ", err)
		return
	}

	switch domainEventDto.EventType {
	case constant.DELETE_UNIT:
		a.handleDeleteUnit(ctx, &domainEventDto)
	default:
		log.Error(ctx, "Unknown event", nil)
	}
}

func (a *AccommodationEventHandler) handleDeleteUnit(ctx context.Context, domainEventDto *basedto.DomainEventDto) {
	var deleteUnitDto request.DeleteUnitDto
	dataBytes, err := json.Marshal(domainEventDto.Data)
	if err != nil {
		log.Error(ctx, "error marshal data from domain event ", err)
		return
	}
	if err := json.Unmarshal(dataBytes, &deleteUnitDto); err != nil {
		log.Error(ctx, "error unmarshal data from domain event ", err)
		return
	}

	if err = a.accService.DeleteUnit(ctx, deleteUnitDto.AccommodationId, deleteUnitDto.UnitId); err != nil {
		log.Error(ctx, "error delete unit ", err)
		return
	}
}

func NewAccommodationEventHandler(accService service.IAccommodationService) *AccommodationEventHandler {
	return &AccommodationEventHandler{accService: accService}
}

File: search_service/src/main/java/huy/project/search_service/ui/kafka/AccommodationEventHandler.java
package huy.project.search_service.ui.kafka;

import huy.project.search_service.core.domain.constant.TopicConstant;
import huy.project.search_service.core.domain.dto.kafka.DomainEventDto;
import huy.project.search_service.core.domain.dto.request.DeleteUnitDto;
import huy.project.search_service.core.service.IAccommodationService;
import huy.project.search_service.kernel.utils.JsonUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.KafkaHandler;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;

@KafkaListener(topics = TopicConstant.AccommodationEvent.TOPIC)
@RequiredArgsConstructor
@Component
@Slf4j
public class AccommodationEventHandler {
    private final IAccommodationService accommodationService;
    private final JsonUtils jsonUtils;


    @KafkaHandler
    public void handleMessage(String message) {
        log.info("Received message: {}", message);
        try {
            var domainEventDto = jsonUtils.fromJson(message, DomainEventDto.class);
            switch (domainEventDto.getEventType()) {
                case TopicConstant.AccommodationEvent.DELETE_UNIT:
                    handleDeleteUnit(domainEventDto.getData());
                    break;
                default:
                    log.error("Unknown event: {}", domainEventDto.getEventType());
            }
        } catch (Exception e) {
            log.error("Failed to handle message: {}, err: {}", message, e.getMessage());
        }
    }

    private void handleDeleteUnit(Object data) {
        String deleteUnitStr = jsonUtils.toJson(data);
        var dto = jsonUtils.fromJson(deleteUnitStr, DeleteUnitDto.class);
        accommodationService.deleteUnit(dto.getAccommodationId(), dto.getUnitId());
    }
}
